tools/cleanup/cleanup.go

```diff
@@ -0,0 +1,308 @@
+		if creation.Before(before) {
+			if *dryRun {
+				log.Printf("[DRY RUN] delete %q", project+"/"+cluster.Name)
+			} else {
+				region, zone := gke.RegionZoneFromLoc(cluster.Location)
+				if err := gkeClient.DeleteCluster(project, region, zone, cluster.Name); err != nil {
+					return count, errors.Wrapf(err, "error deleting cluster %q in project %q", cluster.Name, project)
+				}
+				count++
+			}
+		}
+	}
+	return count, nil
+}
+
+// deleteImages deletes old docker images from a given project.
+func deleteImages(project string, before time.Time) (int, error) {
+	repoRoot := *gcr + "/" + project
+	// TODO(adrcunha): This should be a helper function, like https://github.com/knative/pkg/blob/6d806b998379948bd0107d77bcd831e2bdb4f3cb/testutils/clustermanager/e2e-tests/gke.go#L281
+	if repoRoot == "gcr.io/knative-releases" || repoRoot == "gcr.io/knative-nightly" {
+		return 0, fmt.Errorf("cleaning up %q is forbidden", repoRoot)
+	}
+	gcrrepo, err := name.NewRepository(repoRoot)
+	if err != nil {
+		return 0, errors.Wrapf(err, "cannot open registry %q", repoRoot)
+	}
+	count := 0
+	// Walk down the registry, checking all images and deleting the old ones.
+	return count, google.Walk(gcrrepo, func(repo name.Repository, tags *google.Tags, err error) error {
+		// If we got an error, just return it, there's nothing to do here.
+		if tags == nil {
+			return err
+		}
+		for k, m := range tags.Manifests {
+			ref := repo.String() + "@" + k
+			age := int(time.Since(m.Uploaded).Hours() / 24)
+			log.Printf("%q is %d days old", ref, age)
+			if m.Uploaded.Before(before) {
+				if *dryRun {
+					log.Printf("[DRY RUN] delete %q", ref)
+				} else {
+					// Delete all tags first, otherwise the image can't be deleted.
+					for _, tag := range m.Tags {
+						if err := deleteImage(repo.String() + ":" + tag); err != nil {
+							return err
+						}
+					}
+					if err := deleteImage(ref); err != nil {
+						return err
+					}
+					log.Printf("Deleted %q (uploaded on %s)", ref, m.Uploaded)
+					count++
+				}
+			}
+		}
+		return nil
+	}, google.WithAuthFromKeychain(auther))
+}
+
+// deleteResources call resourceDeleter functions in parallel, one for each given project.
+func deleteResources(projects []string, before time.Time, deleteFunc resourceDeleter) (int, []string) {
+	// Map of errors returned from functions, to avoid repeated entries.
+	var errs = make(map[string]error)
+
+	// Locks for the concurrent tasks.
+	var errsLock sync.RWMutex
+	var countLock sync.RWMutex
+	var wg sync.WaitGroup
+
+	// Blocking channel to keep concurrency under control.
+	concurrencyChan := make(chan struct{}, *concurrentOperations)
+	defer close(concurrencyChan)
+
+	count := 0
+	for i := range projects {
+		wg.Add(1)
+		go func(project string) {
+			defer wg.Done()
+			// Block processing if concurrency reached the limit.
+			concurrencyChan <- struct{}{}
+			// Do not process if previous invocations failed. This prevents a large
```

Comment:
I think there is a race condition here, since this for loop schedule goroutines without any wait time in between, all of the projects probably would start at the same time, and they lined up waiting for `errsLock` then enter the next phase one right after another, so when the first goroutine hit an error some of the other goroutines probably already passed this check. 
How about make it wait for a few seconds before scheduling the next project? Or just wait when `i == 1`?

Reply:
Your suggested wait will throttle much more than the counter lock above. :D I'm not concerned about this case, because the idea is to avoid a flooding of errors, it's OK if a few routines try the resource deletion even if a failure occurred (it won't break the tool or the process).